src/main/java/com/unstoppabledomains/exceptions/NamingServiceException.java

```diff
@@ -0,0 +1,47 @@
+package com.unstoppabledomains.exceptions;
+
+public class NamingServiceException extends Exception {
+  private static final long serialVersionUID = 1L;
+  private NSExceptionCode code;
+  private NSExceptionParams params;
+  private Throwable cause;
+
+  public NamingServiceException(NSExceptionCode code, NSExceptionParams params, Throwable cause) {
+    super(messageFromCode(code, params), cause);
+    this.code = code;
+    this.params = params;
+    this.cause = cause;
```

Comment:
Shouldn't we use `this.initCause` here instead? Also, I am not sure this is even needed because you pass the `cause` into `super`.